watchtower/wtclient/client.go

```diff
@@ -698,101 +680,44 @@ func (c *TowerClient) Stop() error {
-		// pipeline if a ForceQuitDelay is specified. This will have no
-		// effect if the pipeline shuts down cleanly before the delay
-		// fires.
-		//
-		// For full safety, this can be set to 0 and wait out
-		// indefinitely.  However for mobile clients which may have a
-		// limited amount of time to exit before the background process
-		// is killed, this offers a way to ensure the process
-		// terminates.
-		if c.cfg.ForceQuitDelay > 0 {
-			time.AfterFunc(c.cfg.ForceQuitDelay, c.ForceQuit)
-		}
-
-		// 2. Shutdown the backup queue, which will prevent any further
-		// updates from being accepted. In practice, the links should be
-		// shutdown before the client has been stopped, so all updates
-		// would have been added prior.
-		err := c.pipeline.Stop()
-		if err != nil {
-			returnErr = err
-		}
-
-		// 3. Once the backup queue has shutdown, wait for the main
-		// dispatcher to exit. The backup queue will signal it's
-		// completion to the dispatcher, which releases the wait group
-		// after all tasks have been assigned to session queues.
 		close(c.quit)
-		c.wg.Wait()
 
-		// 4. Since all valid tasks have been assigned to session
-		// queues, we no longer need to negotiate sessions.
-		err = c.negotiator.Stop()
+		// 1. Stop the session negotiator.
+		c.negotiator.Stop()
+		err := c.negotiator.Stop()
 		if err != nil {
 			returnErr = err
 		}
 
-		c.log.Debugf("Waiting for active session queues to finish "+
-			"draining, stats: %s", c.stats)
-
-		// 5. Shutdown all active session queues in parallel. These will
-		// exit once all updates have been acked by the watchtower.
+		// 2. Shutdown all active session queues in parallel. These will
+		// exit once all unhandled updates have been replayed to the
+		// task pipeline.
 		c.activeSessions.ApplyAndWait(func(s *sessionQueue) func() {
-			return s.Stop
+			return func() {
+				err := s.Stop()
+				if err != nil {
+					c.log.Errorf("could not stop session "+
+						"queue: %s", s.ID())
+				}
+			}
 		})
 
-		// Skip log if force quitting.
-		select {
-		case <-c.forceQuit:
-			return
-		default:
-		}
-
-		c.log.Debugf("Client successfully stopped, stats: %s", c.stats)
-	})
-
-	return returnErr
-}
-
-// ForceQuit idempotently initiates an unclean shutdown of the watchtower
-// client. This should only be executed if Stop is unable to exit cleanly.
-func (c *TowerClient) ForceQuit() {
-	c.forced.Do(func() {
-		c.log.Infof("Force quitting watchtower client")
-
-		// 1. Shutdown the backup queue, which will prevent any further
-		// updates from being accepted. In practice, the links should be
-		// shutdown before the client has been stopped, so all updates
-		// would have been added prior.
-		err := c.pipeline.Stop()
-		if err != nil {
-			c.log.Errorf("could not stop backup queue: %v", err)
+		// 3. Shutdown the backup queue, which will prevent any further
+		// updates from being accepted.
+		if err := c.pipeline.Stop(); err != nil {
```

Comment:
I'm not sure if maybe this change in order could be the cause for the unit race failure?
I was able to reproduce it locally with `env CGO_ENABLED=1 GORACE="history_size=7 halt_on_errors=1" go test -tags="dev nolog lowscrypt" -test.timeout=3m -test.count=1 -race github.com/lightningnetwork/lnd/watchtower/wtclient` and saw that the clients are blocked on https://github.com/lightningnetwork/lnd/blob/450e3b1011e2f351531d2df25c40481823da83d9/watchtower/wtclient/session_queue.go#L260.
I don't have the full context of how the session/pipeline/client interact, so I could be totally wrong.

Reply:
replied to the other comment :)